javalib/src/main/scala/java/util/HashMap.scala

```diff
@@ -12,153 +12,502 @@
+    if (contentSize + 1 >= threshold)
+      growTable()
+    val idx = index(hash)
+    put0(key, value, hash, idx)
+  }
+
+  /** Adds a key-value pair to this map
+   *
+   *  @param key the key to add
+   *  @param value the value to add
+   *  @param hash the **improved** hashcode of `key` (see computeHash)
+   *  @param idx the index in the `table` corresponding to the `hash`
+   *  @return the old value associated with `key`, or `null` if there was none
+   */
+  private[this] def put0(key: K, value: V, hash: Int, idx: Int): V = {
+    // scalastyle:off return
+    val newNode = table(idx) match {
+      case null =>
+        val newNode = this.newNode(key, hash, value, null, null)
+        table(idx) = newNode
+        newNode
+      case first =>
+        var prev: Node[K, V] = null
+        var n = first
+        while ((n ne null) && n.hash <= hash) {
+          if (n.hash == hash && safeEquals(key, n.key)) {
+            touchNode(n)
+            val old = n.value
+            n.value = value
+            return old
           }
+          prev = n
+          n = n.next
         }
-      }
+        val newNode = this.newNode(key, hash, value, prev, n)
+        if (prev eq null)
+          table(idx) = newNode
+        else
+          prev.next = newNode
+        if (n ne null)
+          n.previous = newNode
+        newNode
     }
+    contentSize += 1
+    nodeWasAdded(newNode)
+    null.asInstanceOf[V]
+    // scalastyle:on return
   }
 
-  private class KeySet extends AbstractSet[K] with AbstractMapView[K] {
-    override def remove(o: Any): Boolean = {
-      val boxedKey = Box(o.asInstanceOf[K])
-      val contains = inner.contains(boxedKey)
-      if (contains)
-        inner -= boxedKey
-      contains
-    }
+  /** Removes a key from this map if it exists.
+   *
+   *  @param key the key to remove
+   *  @return the node that contained `key` if it was present, otherwise null
+   */
+  private def remove0(key: Any): Node[K, V] = {
+    val (node, idx) = findNodeAndIndexForRemoval(key)
+    if (node ne null)
+      remove0(node, idx)
+    node
+  }
 
-    override def iterator(): Iterator[K] = {
-      new AbstractMapViewIterator[K] {
-        protected def getNextForm(key: Box[K]): K =
-          key.inner
+  private[util] final def removeNode(node: Node[K, V]): Unit =
+    remove0(node, index(node.hash))
+
+  private def remove0(node: Node[K, V], idx: Int): Unit = {
+    val previous = node.previous
+    val next = node.next
+    if (previous eq null)
+      table(idx) = next
+    else
+      previous.next = next
+    if (next ne null)
+      next.previous = previous
+    contentSize -= 1
+    nodeWasRemoved(node)
+  }
+
+  /** Grow the size of the table (always times 2). */
+  private[this] def growTable(): Unit = {
+    val oldTable = table
+    val oldlen = oldTable.length
+    val newlen = oldlen * 2
+    threshold = newThreshold(newlen)
+    if (contentSize == 0) {
```

Comment:
Indeed. This cannot happen in this implementation. I took that optimization from `s.c.m.HashMap`, where it can happen because of the `def sizeHint()` method of Scala collections. But `j.u.HashMap` does not have that feature, so it's moot. I'll remove it.